### PR TITLE
[MODERN] Add documentation for granular mutation updates (via optimisticUpdater and updater) (fixes #1652)

### LINKED ISSUE #1652: [Modern] How to use updater and optimisticUpdater
I'm trying to figure out the new cache update mechanisms. In trying to do so, I've hit a few major hurdles (mainly because it's not documented yet).

This is how I got things working:
```
commitMutation(environment, {
  mutation: graphql`
    mutation AppAddTagMutation($input: AddTagInput!) {
        addTag(input: $input) { tags }
    }
    `,
  variables: {
    input: { tag },
  },
  onCompleted: response => console.log(response),
  onError: error => console.error(error),
  updater: store =>
    store.getRoot().copyFieldsFrom(store.getRootField('addTag')),
  optimisticUpdater: store => {
    store
      .getRoot()
      ._mutator.setValue(store.getRoot()._dataID, 'tags', [
        ...tags,
        tag + '<--- optimistic',
      ]);
  },
});
```

When trying to use the store like this:
`store.getRoot().setValue([...tags, tag], 'tags')`, Relay complains with the following invariant: https://github.com/facebook/relay/blob/fcfe1b8496702ff29116445ee2c7b8453377159e/packages/relay-runtime/mutations/RelayRecordProxy.js#L70

Further, trying to poke around, it's not apparent how to get to the response data from within the `updater` or `optimisticUpdater`.

Another note - if I add a `optimisticResponse`, but no `optimisticUpdater`, the UI isn't updated until I receive data from the server. If I add the `optimisticUpdater`, the `optimisticResponse` becomes kind of pointless. I'd expect the `optimisticResponse` to run through the same `updater`, if no `optimisticUpdater` is provided.

Also, I'd expect the same code to work for both `updater` and `optimisticUpdater`. Yet when I have
```
optimisticUpdater: store => store.getRoot().copyFieldsFrom(store.getRootField('addTag'));
```
it produces an error, with and without the `optimisticResponse`.

### PR DESCRIPTION
**Pull request by colepatrickturner**
The purpose of this PR is to add much needed document for Relay Modern. The documentation I have added adds some clarity on how to transverse and mutate the Relay store for advanced mutations. This is necessary for performing equilateral updates optimistically and in response to a mutation. Many developers on Reactiflux and StackOverflow are seeking this information and it would be valuable to have it more readily accessible.

I tried to keep it as low level as possible and relevant. It is my belief that some of the docs have been designed to be more entry and mid level so I've indicated that these APIs are more advanced.

Right now there is a link inside the documentation to the following API
https://github.com/facebook/relay/blob/d0310d69012bba615dacf614319bcf47ee2a0f3f/packages/relay-runtime/ARCHITECTURE.md

If this is appropriate, no further action is necessary, though I believe this document should be exposed, and maybe the documentation I have added should be co-located with Architecture.md. 

Since I have only modified markdown files, testing is not relevant but here are the results from `npm test`.
```
Test Suites 178 passed, 178 total
Tests       2118 passed, 2118 total
Snapshots   12 passed, 12 total
Time        26.834s
Ran all test suites.
```

Lastly I have added an example that's based on the TodoMVC basic example. Please let me know that it is idiomatic or needs reworking. 

Thanks! It's been a pleasure working with Relay Modern.

### CODE REVIEW

File: docs/modern/Mutations.md [hunk 1]
```diff
@@ -221,3 +224,80 @@ const configs = [{
...
+### getLinkedRecord(name: string, args?: ?Variables): ?RecordProxy
+### getLinkedRecords(name: string, args?: ?Variables): ?Array<?RecordProxy>
+Retrieves record(s) associated with the given record, transversing the source by field name and an object representing pre-defined argument values.
+
+
+### setLinkedRecord(record: RecordProxy, name: string, args?: ?Variables): RecordProxy 
+### setLinkedRecords(records: Array<?RecordProxy>, name: string, args?: ?Variables ): RecordProxy
+Updates the records associated with a mutable record, transversing the source by field name and an object representing pre-defined argument values.
+
+### getOrCreateLinkedRecord(name: string, typeName: string, args?: ?Variables ): RecordProx
```
Comment (CarlRosell) [inline docs/modern/Mutations.md]: Great additions to the mutation docs 👍. Seems like there is an `y` missing for the returning flowtype right here.

Comment (colepatrickturner) [inline docs/modern/Mutations.md]: Great catch, thank you! I've rebased and pushed a fix.

**Comment by AndreiCalazans** [comment]
This should be added to the site for easier find.

**Comment by leebyron** [comment]
Thanks!

**Pull request closed without merging**